### PR TITLE
Added new API

### PR DESCRIPTION
**Pull request by zTrap**
## Changes: 
- pumped deps versions 
 - cleaned build script
 - optimized .gitignore
 - code cleanup
 - code style is brought to the [standard](https://google.github.io/styleguide/javaguide.html), provided by Google
 - fixed license file format 

## Additions:
 - added new (un)bind delegations methods
 - added delegate for consuming and managing show/hide events for empty view
 - added travis configuration
 - added missed annotations
 - added new `remove` (by direct item) method
 - added `findViewById` method into holder (also added usage showcase)
 - added `getContext` method into holder (also added usage showcase)
 - added constant which defines the maximum count of undo stack
 - fulled license terms and condition

## Deletions:
 - removed absolutely unnecessary usages of 'final' modifier
 - removed unused resources

### CODE REVIEW

File: example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java [hunk 1]
```diff
@@ -1,25 +1,35 @@
...
 
-    public HeaderViewHolder(final View view) {
+    public HeaderViewHolder(@NonNull View view) {
         super(view);
         ButterKnife.bind(this, view);
     }
 
+    @Override
+    public void bindHolder(MonthItem monthItem) {
+        if (monthItem.getType() == MonthItem.MonthItemType.MONTH) {
```
Comment (thesurix) [inline example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java]: Please check out examples, I can not see headers.

**Review by thesurix — COMMENTED** [review]
Hello

Thank you for contributing to my library. 
I have checked example app and  there is a crash in one of the fragments:

```java
java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid view holder adapter positionViewHolder{967ce90 position=6 id=-1, oldPos=5, pLpos:5 scrap [attachedScrap] tmpDetached no parent} android.support.v7.widget.RecyclerView{930cc58 VFED..... .F....I. 0,0-1020,1692 #7f08007a app:id/recycler_view}, adapter:com.thesurix.example.gesturerecycler.adapter.MonthsAdapter@c1a01b1, layout:android.support.v7.widget.LinearLayoutManager@37cd696, context:com.thesurix.example.gesturerecycler.RecyclerActivity@703dfda
        at android.support.v7.widget.RecyclerView$Recycler.validateViewHolderForOffsetPosition(RecyclerView.java:5610)
        at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5792)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5752)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5748)
        at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2232)
        at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1559)
        at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1519)
        at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:614)
        at android.support.v7.widget.RecyclerView.dispatchLayoutStep1(RecyclerView.java:3763)
        at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3527)
        at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4082)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.support.v7.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java:443)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1743)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1586)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1495)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:336)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:273)
        at com.android.internal.policy.PhoneWindow$DecorView.onLayout(PhoneWindow.java:2678)
        at android.view.View.layout(View.java:16669)
        at android.view.ViewGroup.layout(ViewGroup.java:5440)
        at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2207)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1963)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1139)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6064)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:860)
        at android.view.Choreographer.doCallbacks(Choreographer.java:672)
        at android.view.Choreographer.doFrame(Choreographer.java:608)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:846)
    	at androi
```

Please fix this issue before full PR review.

File: example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java (hunk 1 above)
Comment (zTrap) [reply example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java]: @thesurix Yes, my bad. Fixed this

File: library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java [hunk 2]
```diff
@@ -172,15 +199,15 @@ public void setData(final List<T> data, final DiffUtil.Callback diffCallback) {
      * Clears data.
      */
     public void clearData() {
```
Comment (thesurix) [inline library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java]: I found strange behavior. When I want to clear data and then add data  in `EmptyViewFragment`, there is strange view artifact inside RecyclerView. This issue is not present when we have ` notifyDataSetChanged()` here,

**Comment by zTrap** [comment]
Is any more changes needed?

**Comment by thesurix** [comment]
Have you seen my comment about glitch in `EmptyViewFragment`?

File: library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java (hunk 2 above)
Comment (zTrap) [reply library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java]: reverted

**Comment by zTrap** [comment]
So?

File: library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java [hunk 3]
```diff
@@ -73,21 +73,29 @@
...
     }
 
-    private static final int INVALID_DRAG_POS = -1;
+    /**
+     * The maximum size of undo stack to allocate.
+     * Some VMs reserve some header words in an array.
+     * Attempts to allocate larger arrays may result in
+     * OutOfMemoryError: Requested array size exceeds VM limit
+     * */
+    public final static int MAX_UNDO_STACK_SIZE = Integer.MAX_VALUE - 8;
```
Comment (thesurix) [inline library/src/main/java/com/thesurix/gesturerecycler/GestureAdapter.java]: I don't understand why we need this constant, it is only used in JD. There is no validation etc. Can you give me some information about it?

File: example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java [hunk 4]
```diff
@@ -1,25 +1,33 @@
...
+
 import butterknife.BindView;
 import butterknife.ButterKnife;
 
-
-public class HeaderViewHolder extends GestureViewHolder {
+public class HeaderViewHolder extends GestureViewHolder<MonthItem> {
 
     @BindView(R.id.header_text)
     TextView mHeaderText;
```
Comment (thesurix) [inline example/src/main/java/com/thesurix/example/gesturerecycler/adapter/HeaderViewHolder.java]: Please use`findViewById()` like in `MonthViewHolder.java`

File: library/src/main/java/com/thesurix/gesturerecycler/EmptyViewDataObserver.java [hunk 5]
```diff
@@ -9,41 +10,58 @@
...
     }
 
     @Override
-    public void onItemRangeRemoved(final int positionStart, final int itemCount) {
+    public void onItemRangeRemoved(int positionStart, int itemCount) {
         updateEmptyViewState();
     }
 
-    public void setEmptyView(final View emptyView) {
+    public void setEmptyView(@Nullable View emptyView) {
```
Comment (thesurix) [inline library/src/main/java/com/thesurix/gesturerecycler/EmptyViewDataObserver.java]: This method is not used.

**Comment by zTrap** [comment]
This constant is necessary to avoid `OutOfMemoryError`. It's maximum available array's size

**Comment by thesurix** [comment]
I don't see any reason to create constant only for Java Docs. Another thing is that you made more changes in last PR than requested by review, and again there is a mess with convention. You have ` private @Nullable View mForegroundView` in `MonthViewHolder.java` and `@Nullable private View mEmptyView` in `EmptyViewDataObserver.java`. I am closing this PR. Please make separate PRs for finals, naming convention, API changes etc. Now this PR is big and it is very hard to make review.

**Pull request closed without merging**